### PR TITLE
Sprint 9

### PR DESCRIPTION
**Pull request by leiasouza**
Autenticando o usuário
Autorizando o usuário

### CODE REVIEW

File: account/src/helpers/token.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+//     return response.json([{_id: 1, name: 'Alice'}]);
+
+// });
+// app.post('/login', async (request, response) => {
+//     if(request.body.user === 'Alice'  && request.body.password === 'SenhaDaAlice') {
+//        const token = jwt.sign({ _id: 1 }, compare, {expiresIn: 14400 });
+//         return response.json(201);
+//     }
+//     response.status(400).end();
+//     });
```
Comment (wandersonmaceds) [inline account/src/helpers/token.js:22]: pode remover o código comentado :)

File: account/src/repositories/accountRepository.js [hunk 2]
```diff
@@ -14,4 +14,26 @@ export async function saveAccount(account) {
...
+
+//     return user;
+// }
+
+
+// export async function existsAccountById(id) {
+//     const account = await findAccountById(id);
+
+//     return account !== null;
+// }
```
Comment (wandersonmaceds) [inline account/src/repositories/accountRepository.js:39]: Pode remover :)

**Review by wandersonmaceds — APPROVED** [review]

File: account/src/helpers/token.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+export function generateToken (id) {
+    const tokenSecret = process.env.TOKEN_SECRET;
+    const tokenExpires = process.env.TOKEN_EXPIRATION;
+
+    const token = jwt.sign({ userId: id}, tokenSecret, { expiresIn: `${tokenExpires}s` });
+    return token;
+}
+
+
+// app.get('/user', (request, response) =>{
```
Comment (DPechetti) [inline account/src/helpers/token.js:12]: remove comments

File: account/src/repositories/accountRepository.js [hunk 4]
```diff
@@ -14,4 +14,26 @@ export async function saveAccount(account) {
     await usersCollection.insertOne(account);
     await client.close();
```
Comment (DPechetti) [inline account/src/repositories/accountRepository.js:15]: where are you opening the connection?

File: account/src/repositories/accountRepository.js (hunk 2 above)
Comment (DPechetti) [inline account/src/repositories/accountRepository.js:39]: remove comments

File: account/src/routes.js [hunk 5]
```diff
@@ -15,4 +16,20 @@ router.post('/accounts', async (request, response) => {
...
 
+router.post('/tokens', async (request, response) => {
+   const { email, password } = request.body;
+   const authToken = await createUserTokenUseCase(email, password);
+   
+   if(authToken) {
+        return response.status(201).json({
+            token: authToken
+        });
+   }
```
Comment (DPechetti) [inline account/src/routes.js:27]: i suggest you always put the happy flow at the end

File: account/src/use-case/createUserToken.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    if(!userFound){
+        return null;
+    }
+
+    const passwordIsMatch = await comparePassword(password, userFound.password);
+
+    if (passwordIsMatch) {
+        return generateToken(userFound._id);
+
+    }
```
Comment (DPechetti) [inline account/src/use-case/createUserToken.js:17]: i suggest you always put the happy flow at the end

File: product/src/helpers/token.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+export function decriptToken (token) {
+    try {
+        const tokenSecret = process.env.TOKEN_SECRET;
+        const {userId} = jwt.verify(token, tokenSecret);
+    
+        return {userId};
+    } catch(error){
+        console.log('error when verifying token', error.message);
+        return{};
```
Comment (DPechetti) [inline product/src/helpers/token.js:11]: why are you returning an empty object?

File: product/src/test/data/helpers/products.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import { findProducts } from "../../../repositories/productRepository.js";
+export async function clearProductTable(){
+
+const products = await findProducts();
```
Comment (DPechetti) [inline product/src/test/data/helpers/products.js:4]: fix indentation please :)

File: product/src/test/e2e/createProduct.test.js [hunk 9]
```diff
@@ -1,13 +1,15 @@
...
 
-describe("Criação de produto", () => {
-  afterAll(async() => {
-    await client.close();
-  });
 
+describe("Criação de produto", () => {
+  //afterEach(async() => {
+    //await clearProductTable();
+  //});
```
Comment (DPechetti) [inline product/src/test/e2e/createProduct.test.js:11]: remove comments

File: account/src/helpers/token.js (hunk 1 above)
Comment (leiasouza) [reply account/src/helpers/token.js]: Corrigido.

File: account/src/repositories/accountRepository.js (hunk 2 above)
Comment (leiasouza) [reply account/src/repositories/accountRepository.js]: Corrigido.

File: account/src/helpers/token.js (hunk 3 above)
Comment (leiasouza) [reply account/src/helpers/token.js]: adjusted.

File: account/src/repositories/accountRepository.js (hunk 2 above)
Comment (leiasouza) [reply account/src/repositories/accountRepository.js]: adjusted.

File: product/src/test/data/helpers/products.js (hunk 8 above)
Comment (leiasouza) [reply product/src/test/data/helpers/products.js]: adjusted.

File: product/src/test/e2e/createProduct.test.js (hunk 9 above)
Comment (leiasouza) [reply product/src/test/e2e/createProduct.test.js]: adjusted.

File: account/src/repositories/accountRepository.js [hunk 10]
```diff
@@ -12,7 +12,6 @@ export async function saveAccount(account) {
     await client.connect();
     const usersCollection = await getUsersCollection(client);
     await usersCollection.insertOne(account);
-    await client.close();
```
Comment (DPechetti) [inline account/src/repositories/accountRepository.js:15]: i think if you remove the close it's possible to open multiple connections at the same time

File: account/src/routes.js (hunk 5 above)
Comment (leiasouza) [reply account/src/routes.js]: adjusted.

File: account/src/use-case/createUserToken.js (hunk 6 above)
Comment (leiasouza) [reply account/src/use-case/createUserToken.js:17]: adjusted.

File: account/src/repositories/accountRepository.js (hunk 10 above)
Comment (leiasouza) [reply account/src/repositories/accountRepository.js:15]: adjusted.

**Review by DPechetti — APPROVED** [review]

**Pull request merged**